design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,204 @@
+is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### Protobuf
+Protobuf 3 has a Timestamp type. The Protouf deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be deprecated:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
+
+These functions will still be available so that existing queries using these functions won't break,
+but the documentation will state that they are deprecated and will direct users towards using the
+TIMESTAMP type and the new functions.
+
+The following functions will be added:
+* `PARSE_TIMESTAMP(format, timestamp_string)` - converts a string into a TIMESTAMP 
+* `FORMAT_TIMESTAMP(format, timestamp)` - returns a string in the specified format
+* `NOW()` - returns the time after the issuing query is done executing
+* `CONVERT_TZ(timestamp, from_tz ,to_tz)` - converts a timestamp from one timezone to another
+
+There are a few existing UDFs that deal with dates. These should be left as is until a DATE type is implemented:
+
+* UNIX_DATE
+* DATETOSTRING
+* STRINGTODATE
+
+### Casting
+
+Casting from TIMESTAMP to STRING will return the timestamp in ODBC canonical form with millisecond
+precision (yyyy-mm-dd HH:mm:ss:fff), and  casting from STRING to TIMESTAMP will attempt to parse the
+string into a TIMESTAMP.
+
+### Arithmetic operations and comparisons
+
+At the very least, we would want KSQL to be able to compare, add and subtract TIMESTAMPS.
+
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html) and [MariaDB](https://mariadb.com/kb/en/date-time-functions/)
+have very complete lists of built-in arithmetic functions, while [PostgreSQL](https://www.postgresql.org/docs/9.0/functions-datetime.html)
+defines behaviors for arithmetic operators.
+
+Because functions are more well-defined, we should opt for built-in arithmetic functions
+instead of using operators. The following functions are necessary:
+
+* `TIMESTAMP_ADD(time_stamp, duration)` - adds a duration to `time_stamp` and returns the result as a
+TIMESTAMP
+* `TIMESTAMP_SUB(time_stamp1, duration)` - subtracts a duration from `time_stamp` and returns the
+result as a TIMESTAMP
+
+Durations will be expressed in the form, `<integer_value> <unit>`. This is discussed further in the
+next section.
+
+As for comparisons, the following expressions should be supported:
+```
+time_stamp1 < time_stamp2
+time_stamp1 > time_stamp2
+time_stamp1 = time_stamp2
+time_stamp1 BETWEEN time_stamp2 AND time_stamp3
+```
+
+Comparisons between TIMESTAMPs and other data types should not be allowed.
+
+### Durations / Time units
```

Comment:
What units will we support? Only DAY?

Could you add a list of units to support?